File: 173-binary-search-tree-iterator/binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    Stack<TreeNode> s;
    public void put(TreeNode r){
        TreeNode ptr=r;
        while(ptr.left!=null){
           s.push(ptr.left);
           ptr=ptr.left;
        }
    }
    public BSTIterator(TreeNode root) {
        s=new Stack<>();
        s.push(root);
        put(root);
    }
    
    public int next() {
        if(s.size()==0) return -1;
        TreeNode temp=s.pop();
        if(temp.right!=null){
           s.push(temp.right);
           put(temp.right);
        }
        int c=temp.val;
        return c;
    }
    
    public boolean hasNext() {
        if(s.size()!=0) return true;
        return false;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 173-binary-search-tree-iterator/binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.root = root
        self.stack = []
        ptr = root
        while ptr != None:
            self.stack.append(ptr)
            ptr = ptr.left
        

    def next(self):
        """
        :rtype: int
        """
        ptr = self.stack[-1]
        self.stack.pop(-1)
        c = ptr.val
        if ptr.right != None:
            ptr = ptr.right
            while ptr != None:
                self.stack.append(ptr)
                ptr = ptr.left
        
        return c
        

    def hasNext(self):
        """
        :rtype: bool
        """
        if len(self.stack)!= 0:
            return True
        return False
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()